hazelcast/src/main/java/com/hazelcast/ringbuffer/impl/operations/MergeOperation.java

```diff
@@ -25,74 +25,107 @@
+public class MergeOperation extends Operation
+        implements IdentifiedDataSerializable, BackupAwareOperation, ServiceNamespaceAware {
 
     private ObjectNamespace namespace;
-    private SplitBrainMergePolicy<Object, RingbufferMergeTypes> mergePolicy;
-    private List<RingbufferMergeTypes> mergingEntries;
+    private SplitBrainMergePolicy<RingbufferMergeData<Object>, RingbufferMergeTypes> mergePolicy;
+    private Ringbuffer<Object> mergingRingbuffer;
 
+    private transient Ringbuffer<Object> resultRingbuffer;
     private transient RingbufferConfig config;
-    private transient RingbufferContainer<Object, Object> ringbuffer;
-    private transient Map<Long, Data> valueMap;
 
     public MergeOperation() {
     }
 
-    public MergeOperation(ObjectNamespace namespace, SplitBrainMergePolicy<Object, RingbufferMergeTypes> mergePolicy,
-                          List<RingbufferMergeTypes> mergingEntries) {
+    public MergeOperation(ObjectNamespace namespace,
+                          SplitBrainMergePolicy<RingbufferMergeData<Object>, RingbufferMergeTypes> mergePolicy,
+                          Ringbuffer<Object> mergingRingbuffer) {
         this.namespace = namespace;
         this.mergePolicy = mergePolicy;
-        this.mergingEntries = mergingEntries;
+        this.mergingRingbuffer = mergingRingbuffer;
     }
 
     @Override
     public void beforeRun() throws Exception {
-        RingbufferService service = getService();
+        final RingbufferService service = getService();
         config = getRingbufferConfig(service, namespace);
-        ringbuffer = service.getOrCreateContainer(getPartitionId(), namespace, config);
     }
 
     @Override
     public void run() throws Exception {
-        valueMap = createHashMap(mergingEntries.size());
-        for (RingbufferMergeTypes mergingEntry : mergingEntries) {
-            long resultSequence = ringbuffer.merge(mergingEntry, mergePolicy);
-            if (resultSequence != -1) {
-                valueMap.put(resultSequence, ringbuffer.readAsData(resultSequence));
+        final RingbufferService service = getService();
+        final SerializationService serializationService = getNodeEngine().getSerializationService();
+        RingbufferContainer<Object, Object> existingContainer = service.getContainerOrNull(getPartitionId(), namespace);
+
+        final RingbufferMergeTypes mergingValue = getMergingType(mergingRingbuffer);
+        final RingbufferMergeTypes existingValue =
+                existingContainer != null ? getMergingType(existingContainer.getRingbuffer()) : null;
+        serializationService.getManagedContext().initialize(mergePolicy);
+
+        final RingbufferMergeData<Object> resultData = mergePolicy.merge(mergingValue, existingValue);
+
+        if (resultData == null) {
+            service.destroyContainer(getPartitionId(), namespace);
```

Comment:
I see that `destroyDistributedObject` releases other resources too, such as listeners. Don't we need that here? (similar for backup OP)

Reply:
I'm just in the process of changing this and getting it in line with the [queue merge implementation](https://github.com/hazelcast/hazelcast/pull/12802/files#diff-05b88fcce532485445fb1aa3157a57abR1128). OTOH I don't like that the `destroyDistributedObject` is called. What if the merge policy wants to keep some partitions but drop some other partitions?

Reply:
:+1: lets revisit once changed

Reply:
Hm, this was a brain fart - there is only one ringbuffer on one partition (unless it's an event journal ringbuffer) and if the user returns `null` we can simply destroy the object.